src/coreclr/src/jit/lower.cpp

```diff
@@ -3691,10 +3691,17 @@ GenTree* Lowering::LowerDirectCall(GenTreeCall* call)
             // Non-virtual direct calls to addresses accessed by
             // a double indirection.
             //
-            // Double-indirection. Load the address into a register
-            // and call indirectly through the register
+
+            // Expanding an IAT_PPVALUE here, will lose the opportunity
+            // to Hoist/CSE the first indirection as it is an invariant load
+            //
+            assert(!"IAT_PPVALUE case in LowerDirectCall");
```

Comment:
Question: How do we usually check such JIT expectations about VM return values? The current VM implementation (`CEEInfo::getFunctionEntryPoint`, `ZapInfo::getFunctionEntryPoint`) does not use `IAT_PPVALUE`, would it make sense to assert that `pResult->accessType != IAT_PPVALUE` in the VM code before we exit these functions and change this block to a `unreached()`?

Reply:
I don't believe that there is a hard cnad fast contract in this area.

The JIT is free to implement all possible variants and the VM is free to only utilize one or two and never use the third choice.  When I noticed that waiting all the way until lower to materialize these indirections would not be optimal for codegen, I added the assert to find out it it was being used.   It turns out in one case it was used and in another case it wasn't.

I don;t think that this deserves an assert in the VM for this behavior.  I like to reserve asserts for things that are actually incorrect in some semantic way.  In any case if someone changes the VM and nowe asks for a doubel indirection we will get an assert in the JIT which is fine, because that is the component that has an issue with the double indirection case.